### PR TITLE
CRUD web apps:  i18n

### PR DESCRIPTION
**Pull request by wg102**
This PR add i18n to jupyter+volumes using Angular ngx translate

Related to https://github.com/kubeflow/kubeflow/issues/5858
Closes https://github.com/StatCan/kubeflow/issues/56

To test the feature:
- Build the environment as usual
- Modify your browser's language. The code detects the top-most language if multiple are set.
- Languages supported are : English and French.
- It defaults to English if neither of those are the top ones.

### CODE REVIEW

**Comment by saffaalvi** [comment]
@googlebot I consent.

**Comment by wg102** [comment]
Will need to update and rebase once angular version is changed (https://github.com/kubeflow/kubeflow/pull/5805).

**Comment by thesuperzapper** [comment]
@wg102 love this work!

If possible, to help future developers  implement/update localisations, can you add some docs to the README for each of the apps:
* https://github.com/kubeflow/kubeflow/tree/master/components/crud-web-apps/common
* https://github.com/kubeflow/kubeflow/blob/master/components/crud-web-apps/jupyter
* https://github.com/kubeflow/kubeflow/tree/master/components/crud-web-apps/tensorboards
* https://github.com/kubeflow/kubeflow/tree/master/components/crud-web-apps/volumes

The main things we need in those README's are:
* Information about where to find the localisation JSON files (and how to update them)
* Information (generally) about how localisation has been implemented

**Comment by wg102** [comment]
@thesuperzapper I will add those.
I was actually wondering if the en.json (and any other languages) should be present in all of the apps, or once in the common, and get copied to the other ones. Thoughts?

**Comment by DavidSpek** [comment]
@wg102 That's a good question. If there are a lot of common values between the web apps it does make sense to place then JSON files in common. However, it also might confuse or get forgotten by developers as they will need to change files in a different code base than the one they are working on.

**Comment by DavidSpek** [comment]
Here is the log output of the failed tests:

**Jupyter Web App**:

```
[warn] src/app/pages/form/form-default/form-affinity-tolerations/form-affinity-tolerations.component.html
[warn] src/app/pages/form/form-default/form-configurations/form-configurations.component.html
[warn] src/app/pages/form/form-default/form-cpu-ram/form-cpu-ram.component.html
[warn] src/app/pages/form/form-default/form-data-volumes/form-data-volumes.component.html
[warn] src/app/pages/form/form-default/form-default.component.ts
[warn] src/app/pages/form/form-default/form-default.module.ts
[warn] src/app/pages/form/form-default/form-gpus/form-gpus.component.ts
[warn] src/app/pages/form/form-default/form-image/form-image.component.html
[warn] src/app/pages/form/form-default/form-workspace-volume/form-workspace-volume.component.ts
[warn] src/app/pages/form/form-rok/form-rok.component.ts
[warn] src/app/pages/index/index-default/index-default.component.ts
[warn] Code style issues found in the above file(s). Forgot to run Prettier?
-----------------------------------------------
Please run `npm run format:write` to format your code.
-----------------------------------------------
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! frontend@0.0.0 format:check: `prettier --check 'src/**/*.{js,ts,html,scss,css}' || node scripts/check-format-error.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the frontend@0.0.0 format:check script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2021-06-04T10_14_14_347Z-debug.log
```

**Volumes Web App (backend)**:

```
W
Collecting flake8
  Downloading flake8-3.9.2-py2.py3-none-any.whl (73 kB)
Collecting importlib-metadata
  Downloading importlib_metadata-4.5.0-py3-none-any.whl (17 kB)
Collecting pycodestyle<2.8.0,>=2.7.0
  Downloading pycodestyle-2.7.0-py2.py3-none-any.whl (41 kB)
Collecting mccabe<0.7.0,>=0.6.0
  Downloading mccabe-0.6.1-py2.py3-none-any.whl (8.6 kB)
Collecting pyflakes<2.4.0,>=2.3.0
  Downloading pyflakes-2.3.1-py2.py3-none-any.whl (68 kB)
Collecting typing-extensions>=3.6.4
  Downloading typing_extensions-3.10.0.0-py3-none-any.whl (26 kB)
Collecting zipp>=0.5
  Downloading zipp-3.4.1-py3-none-any.whl (5.2 kB)
Installing collected packages: zipp, typing-extensions, pyflakes, pycodestyle, mccabe, importlib-metadata, flake8
Successfully installed flake8-3.9.2 importlib-metadata-4.5.0 mccabe-0.6.1 pycodestyle-2.7.0 pyflakes-2.3.1 typing-extensions-3.10.0.0 zipp-3.4
.1
WARNING: Running pip as root will break packages and permissions. You should install packages reliably by using venv: https://pip.pypa.io/warn
ings/venv
./apps/common/status.py:10:80: E501 line too long (134 > 79 characters)
./apps/common/status.py:14:80: E501 line too long (131 > 79 characters)
./apps/common/status.py:27:80: E501 line too long (138 > 79 characters)
./apps/common/status.py:48:80: E501 line too long (122 > 79 characters)
```

**Common Frontend**:

```
        error properties: Object({ ngTempTokenPath: null, ngTokenPath: [ 'TranslatePipe', Function ], ngDebugContext: DebugContext_({ view: Ob
ject({ def: Object({ factory: Function, nodeFlags: 914344979, rootNodeFlags: 1, nodeMatchedQueries: 30, flags: 0, nodes: [ Object({ nodeIndex:
 0, parent: null, renderParent: null, bindingIndex: 0, outputIndex: 0, checkIndex: 0, flags: 1, childFlags: 914344979, directChildFlags: 33554
433, childMatchedQueries: 30, matchedQueries: Object({  }), matchedQueryIds: 0, references: Object({  }), ngContentIndex: null, childCount: 26
, bindings: [  ], bindingFlags: 0, outputs: [  ], element: Object({ ns: '', name: 'div', attrs: [ Array ], template: null, componentProvider:
null, componentView: null, componentRendererType: null, publicProviders: null({  }), allProviders: null({  }), handleEvent: Function }), provi
der: null, text: null, query: null, ngContent: null }), Object({ nodeIndex: 1, parent: Object({ nodeIndex: 0, parent: null, renderParent: null
, bindingIndex: 0, outputIndex:  ...
            at <Jasmine>
            at NullInjector.get (http://localhost:9876/mnt/test-data-volume/kubeflow-kubeflow-presubmit-common-ui-5880-04ebc5e-4976-48a7/src/k
ubeflow/kubeflow/components/crud-web-apps/common/frontend/kubeflow-common-lib/node_modules/@angular/core/fesm2015/core.js:855:1)
            at resolveToken (http://localhost:9876/mnt/test-data-volume/kubeflow-kubeflow-presubmit-common-ui-5880-04ebc5e-4976-48a7/src/kubef
low/kubeflow/components/crud-web-apps/common/frontend/kubeflow-common-lib/node_modules/@angular/core/fesm2015/core.js:17514:1)
            at tryResolveToken (http://localhost:9876/mnt/test-data-volume/kubeflow-kubeflow-presubmit-common-ui-5880-04ebc5e-4976-48a7/src/ku
beflow/kubeflow/components/crud-web-apps/common/frontend/kubeflow-common-lib/node_modules/@angular/core/fesm2015/core.js:17440:1)
            at StaticInjector.get (http://localhost:9876/mnt/test-data-volume/kubeflow-kubeflow-presubmit-common-ui-5880-04ebc5e-4976-48a7/src
/kubeflow/kubeflow/components/crud-web-apps/common/frontend/kubeflow-common-lib/node_modules/@angular/core/fesm2015/core.js:17266:1)
            at resolveToken (http://localhost:9876/mnt/test-data-volume/kubeflow-kubeflow-presubmit-common-ui-5880-04ebc5e-4976-48a7/src/kubef
low/kubeflow/components/crud-web-apps/common/frontend/kubeflow-common-lib/node_modules/@angular/core/fesm2015/core.js:17514:1)
            at tryResolveToken (http://localhost:9876/mnt/test-data-volume/kubeflow-kubeflow-presubmit-common-ui-5880-04ebc5e-4976-48a7/src/ku
beflow/kubeflow/components/crud-web-apps/common/frontend/kubeflow-common-lib/node_modules/@angular/core/fesm2015/core.js:17440:1)
            at StaticInjector.get (http://localhost:9876/mnt/test-data-volume/kubeflow-kubeflow-presubmit-common-ui-5880-04ebc5e-4976-48a7/src
/kubeflow/kubeflow/components/crud-web-apps/common/frontend/kubeflow-common-lib/node_modules/@angular/core/fesm2015/core.js:17266:1)
            at resolveNgModuleDep (http://localhost:9876/mnt/test-data-volume/kubeflow-kubeflow-presubmit-common-ui-5880-04ebc5e-4976-48a7/src
/kubeflow/kubeflow/components/crud-web-apps/common/frontend/kubeflow-common-lib/node_modules/@angular/core/fesm2015/core.js:30393:1)
..........
HeadlessChrome 88.0.4324 (Linux 0.0.0): Executed 34 of 34 (11 FAILED) (3.524 secs / 3.388 secs)
TOTAL: 11 FAILED, 23 SUCCESS
TOTAL: 11 FAILED, 23 SUCCESS
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! angular-frontend-kubeflow@0.0.0 test-ci: `ng test --no-watch --no-progress --browsers=ChromeHeadlessCI`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the angular-frontend-kubeflow@0.0.0 test-ci script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm ERR! A complete log of this run can be found in:
npm ERR!     /home/blessuser/.npm/_logs/2021-06-04T10_13_48_146Z-debug.log
```

File: components/crud-web-apps/volumes/README.md [hunk 1]
```diff
@@ -40,3 +40,31 @@ make -C backend install-deps
...
+
+For each language added, `app.component.ts` will need to be updated.
+
+**When a language is added:** 
+- Copy the en.json file and rename is to the language you want to add. As it currently is, the culture should not be included.
+- Change the values to the translated ones
+
+**When a language is added:** 
+- Copy the en.json file and rename is to the language you want to add. As it currently is, the culture should not be included.
+- Change the values to the translated ones
```
Comment (DavidSpek) [inline components/crud-web-apps/volumes/README.md:61]: It looks like this section is exactly the same as the one above. Was it meant to be different, or is it a copy/paste error?

Comment (wg102) [reply components/crud-web-apps/volumes/README.md:61]: Oh, yes copy/paste error... thanks for noticing!

**Comment by DavidSpek** [comment]
@Ito-Matsuda You will need to sign the Google CLA and then consent so this PR can move forward.

**Comment by DavidSpek** [comment]
@wg102 @Ito-Matsuda Here is the log output of the failed common UI tests. Looks like a lot of the errors that were previously seen are now gone.

```
HeadlessChrome 88.0.4324 (Linux 0.0.0): Executed 20 of 34 (1 FAILED) (0 secs / 1.744 secs)
HeadlessChrome 88.0.4324 (Linux 0.0.0): Executed 25 of 34 (1 FAILED) (0 secs / 2.199 secs)
WARN: 'The "longpress" event cannot be bound because Hammer.JS is not loaded and no custom loader has been specified.'
HeadlessChrome 88.0.4324 (Linux 0.0.0): Executed 26 of 34 (1 FAILED) (0 secs / 2.237 secs)
WARN: 'The "longpress" event cannot be bound because Hammer.JS is not loaded and no custom loader has been specified.'
HeadlessChrome 88.0.4324 (Linux 0.0.0): Executed 28 of 34 (1 FAILED) (0 secs / 2.355 secs)
WARN: 'The "longpress" event cannot be bound because Hammer.JS is not loaded and no custom loader has been specified.'
HeadlessChrome 88.0.4324 (Linux 0.0.0): Executed 29 of 34 (1 FAILED) (0 secs / 2.378 secs)
WARN: 'The "longpress" event cannot be bound because Hammer.JS is not loaded and no custom loader has been specified.'
HeadlessChrome 88.0.4324 (Linux 0.0.0): Executed 30 of 34 (1 FAILED) (0 secs / 2.4 secs)
WARN: 'The "longpress" event cannot be bound because Hammer.JS is not loaded and no custom loader has been specified.'
HeadlessChrome 88.0.4324 (Linux 0.0.0): Executed 33 of 34 (1 FAILED) (0 secs / 2.445 secs)
HeadlessChrome 88.0.4324 (Linux 0.0.0): Executed 34 of 34 (1 FAILED) (2.646 secs / 2.536 secs)
TOTAL: 1 FAILED, 33 SUCCESS
TOTAL: 1 FAILED, 33 SUCCESS
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! angular-frontend-kubeflow@0.0.0 test-ci: `ng test --no-watch --no-progress --browsers=ChromeHeadlessCI`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the angular-frontend-kubeflow@0.0.0 test-ci script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm ERR! A complete log of this run can be found in:
npm ERR!     /home/blessuser/.npm/_logs/2021-06-11T16_07_20_298Z-debug.log
```

**Comment by Jose-Matsuda** [comment]
@googlebot I fixed it.

**Comment by Jose-Matsuda** [comment]
@googlebot I consent.

**Comment by DavidSpek** [comment]
The Kaniko build of the Volumes Web App just completed successfully, so that doesn't need any fixing.

**Comment by DavidSpek** [comment]
@Jose-Matsuda Thanks for fixing that so quickly. The tests that failed now passed successfully. All that needs to be done is fixing the merge conflicts. A rebase on the current master is probably the best way to solve this.

File: components/crud-web-apps/jupyter/frontend/src/app/app.component.ts [hunk 2]
```diff
@@ -1,10 +1,16 @@
...
 
 @Component({
   selector: 'app-root',
   templateUrl: './app.component.html',
   styleUrls: ['./app.component.scss'],
 })
 export class AppComponent {
+  constructor(private translate: TranslateService) {
+    const currentLanguage = this.translate.getBrowserLang();
+    const lang = currentLanguage.match(/en|fr/) ? currentLanguage : 'en';
```
Comment (DavidSpek) [inline components/crud-web-apps/jupyter/frontend/src/app/app.component.ts:12]: Will new languages need to be added to the `currentLanguage.match(/en|fr/)` section once the JSON files are added?

File: components/crud-web-apps/jupyter/frontend/src/app/pages/index/index-default/index-default.component.ts [hunk 3]
```diff
@@ -52,6 +54,7 @@ export class IndexDefaultComponent implements OnInit, OnDestroy {
     // Poll for new data and reset the poller if different data is found
     this.subs.add(
       this.poller.start().subscribe(() => {
+        //this.config = defaultConfig();
```
Comment (DavidSpek) [inline components/crud-web-apps/jupyter/frontend/src/app/pages/index/index-default/index-default.component.ts:57]: What was this used for? Can it be removed?

**Review by DavidSpek — COMMENTED** [review]
Awesome work everybody! I just have a few small questions, but everything looks good. I've tested the web apps and they all perform as expected and follow the language of the browser.

/lgtm

If one of the @kubeflow/wg-notebooks-leads can look through this we can hopefully get this approved soon so I can proceed with the [Angular Update PR](https://github.com/kubeflow/kubeflow/pull/5805).

File: components/crud-web-apps/jupyter/frontend/src/app/pages/index/index-default/index-default.component.ts (hunk 3 above)
Comment (Jose-Matsuda) [reply components/crud-web-apps/jupyter/frontend/src/app/pages/index/index-default/index-default.component.ts]: Removed.

File: components/crud-web-apps/jupyter/frontend/src/app/app.component.ts (hunk 2 above)
Comment (Jose-Matsuda) [reply components/crud-web-apps/jupyter/frontend/src/app/app.component.ts:12]: That is correct. Once a language's JSON file is added the corresponding `app.component.ts` will need to be updated as well

**Comment by DavidSpek** [comment]
@Jose-Matsuda Thanks for the clarification.

/lgtm

**Comment by thesuperzapper** [comment]
@wg102 after testing this locally, and reviewing the code, it looks good to go, we are incredibly grateful for your hard work on this!

I would say next steps are to consider a language dropdown box in the Central Dashboard (under some kind of user settings). However, be aware that we are considering rewriting Central Dashboard in Angular (https://github.com/kubeflow/kubeflow/issues/6000), so doing anything too drastic would be a waste of effort.

/approve

**Pull request merged**